SYNC.md

```diff
@@ -0,0 +1,143 @@
+      sendMessage({ source: workerId, target: source, type: 'HELLO' })
+    }
+    return
+  }
+```
+
+### Synchronizing with one or more peers
+
+In general, whenever a peer creates a local change or receives a sync message from another peer, it should respond to all the peers it is connected to with its updated status. This will both confirm receipt of any data to the sending peer and also allow other peers to request any changes they may still need. 
+
+Generating new sync messages to other peers is straightforward. Simply call `generateSyncMessage` and, if `syncMessage` is not null, send it to the appropriate peer. You will also need to hold on to the returned `syncState` for that peer, since it keeps track of what data you have sent them to avoid sending data twice.
+
+Here is a simple example:
+```
+function updatePeers(docId: string) {
+  Object.entries(syncStates).forEach(([peer, syncState]) => {
+    const [nextSyncState, syncMessage] = Backend.generateSyncMessage(
+      backends[docId],
+      syncState[docId] || Backend.initSyncState(),
+    )
+    syncStates[peer] = { ...syncStates[peer], [docId]: nextSyncState }
+    if (syncMessage) {
+      sendMessage({
+        docId, source: workerId, target: peer, syncMessage,
+      })
+    }
+  })
+}
+```
+### Receiving sync messages
+Receiving sync messages is also simple. Just pass the document, syncState, and incoming message to `receiveSyncMessage()`, and keep the updated results. Calling `receiveSyncMessage` may also produce a `patch`, which you must forward to the frontend.
+
+After receiving a sync message, you should check if you need to send new sync messages to any connected peers using the code above. In our example code below this is represented by a call to `updatePeers()`
+
+```
+  const [nextBackend, nextSyncState, patch] = Backend.receiveSyncMessage(
+    backends[docId],
+    syncStates[source][docId] || Backend.initSyncState(),
+    syncMessage,
+  )
+  backends[docId] = nextBackend
+  syncStates[source] = { ...syncStates[source], [docId]: nextSyncState }
+
+  updatePeers(docId)
+
+  if (patch) {
+    sendPatchToFrontend({ docId, patch })
+  }
+}
+```
+
+### Applying and distributing local changes
+
+When you create a local change to a document, simply call `generateSyncMessage()` for each peer to produce a message to send them. In general, you can use the same `updatePeers()` implementation for both receiving messages or creating local changes. You may want to rate limit or debounce these communications to reduce network traffic, but this isn't required. *Remember, after applying a local change to the backend you will need to forward the resulting patch to your frontend!*
+
+Here's a sample implementation:
+
+```
+
+// Respond to messages from the frontend document
+self.addEventListener('message', (evt: any) => {
+  const { data } = evt
+  const { docId } = data
+
+  if (data.type === 'OPEN') {
```

Comment:
This `data` object doesn't seem to be defined anywhere, not sure what it can hold. Maybe worth a mention earlier?